### ISSUE TITLE
Converting .wav  file to text in c# using Google cloud speech to text 

### ISSUE THREAD

**Issue by harsh-tukadiya** [googleapis/google-cloud-dotnet#4409]
Converting .wav  file to text in c# using Google cloud speech to text.
I want to transcribe audio(recording) file which is longer than one minute.
I tried for shorter than one minute it's work properly.

Error is :-Grpc.Core.RpcException: 'Status(StatusCode=InvalidArgument, Detail="WAV header indicates an unsupported format.")'

myController.cs

```csharp
                 public HttpResponseMessage Speech() {
                  string storageUri = "gs://bucket/my file.wav";
                var response1 = new HttpResponseMessage(HttpStatusCode.OK);
                var speech = SpeechClient.Create();
                var speakerDiarizationConfig = new SpeakerDiarizationConfig() {
                    EnableSpeakerDiarization = true,
                    
                };
                var longOperation = speech.LongRunningRecognize(new RecognitionConfig() {
                    Encoding = RecognitionConfig.Types.AudioEncoding.Linear16,
                    SampleRateHertz = 8000,
                    AudioChannelCount = 2,
                    EnableWordTimeOffsets = true,
                    LanguageCode = LanguageCodes.English.UnitedKingdom,
                    EnableAutomaticPunctuation = true,                   
                    Model = "phone_call",
                },
                RecognitionAudio.FromStorageUri(storageUri));
                longOperation = longOperation.PollUntilCompleted();
                var response = longOperation.Result;
                string data = null;
                foreach (var result in response.Results) {
                    foreach (var alternative in result.Alternatives) {
                        data = data + alternative.Transcript;
                        Console.WriteLine($"Transcript: { alternative.Transcript}");
                    }
                }
                response1.Content = new StringContent(JsonConvert.SerializeObject(data), encoding: 
                 Encoding.UTF8, "application/json");
                return response1;
            }
```

**Comment by jskeet** [comment]
Are you able to make the wave file public? It's hard to help diagnose the problem unless I can reproduce it.

**Comment by harsh-tukadiya** [comment]
Sorry , I didn't got your point. When I try to convert a .wav file which is longer than 1 minute from speech to text, I get following error

**Error is :-Grpc.Core.RpcException: 'Status(StatusCode=InvalidArgument, Detail="WAV header indicates an unsupported format.")'**

Any way how to resolve it?

**Comment by jskeet** [comment]
My point is that we can see your code, but we don't have a sample data file that's failing for you. If you could make one of those files public, I could try to reproduce your problem with your code and your data. That's likely to be a lot more productive than me creating a wav file and uploading it myself - because if the problem is in the wav file (e.g. it's not the frequency your code says) then I won't be able to reproduce the problem.

**Comment by harsh-tukadiya** [comment]
This is my wav file.
[my file.zip](https://github.com/googleapis/google-cloud-dotnet/files/4248498/my.file.zip)

**Comment by jskeet** [comment]
Testing with that file, I initially receive an error like this:

> Unhandled Exception: Grpc.Core.RpcException: Status(StatusCode=InvalidArgument, Detail="sample_rate_hertz (8000) in RecognitionConfig must either be omitted or match the value in the WAV header ( 44100).")

... which makes sense as the WAV file has a rate of 44.1KHz. Having changed the code to specify that, it succeeds.

I suspect this isn't the file that produced the error you've reported, which means it doesn't help us work out what's going wrong with your actual file :(

**Comment by harsh-tukadiya** [comment]
@jskeet 
For actual file i tried speech to text with
**SampleRateHertz = 44100,
AudioChannelCount = 2,**

I get the following exception
 **Grpc.Core.RpcException: 'Status(StatusCode=InvalidArgument, Detail="sample_rate_hertz (44100) in RecognitionConfig must either be omitted or match the value in the WAV header ( 8000).")'**

After that i change the code with 
 **var longOperation = speech.LongRunningRecognize(new RecognitionConfig() {
Encoding = RecognitionConfig.Types.AudioEncoding.Linear16,
SampleRateHertz = 8000,
AudioChannelCount = 2,**

then i get this error .
**Grpc.Core.RpcException: 'Status(StatusCode=InvalidArgument, Detail="Must use single channel (mono) audio, but WAV header indicates 1 channels.")'**

After this a change channel count  to  1.
I get following error
**Grpc.Core.RpcException: 'Status(StatusCode=InvalidArgument, Detail="WAV header indicates an unsupported format.")'
**
My wav file is confidential so, I am not able to show you.

**Comment by jskeet** [comment]
> My wav file is confidential so, I am not able to show you.

It doesn't need to be the *exact* wav file you're using - just one that's *representative* of the file causing the problem. I would expect you to be able to record a demonstration file in the same way as your real files are recorded, but without any confidential information in. That should demonstrate the same problem.

(To be honest I think it's *highly* unlikely that the problem is the client library here - just as with your previous question, this is really about the service rather than the library. But it's unlikely that you'd be able to get help in other support forums either without being able to provide a sample of the kind of file that causes the problem.)

**Comment by harsh-tukadiya** [comment]
@jskeet 
This is my Original file 
[g.zip](https://github.com/googleapis/google-cloud-dotnet/files/4249697/g.zip)

**Comment by jskeet** [comment]
Right - it looks like that file is compressed using Intel IMA/DVI-ADPCM, which isn't supported by the Cloud Speech API.

See https://cloud.google.com/speech-to-text/docs/encoding for details of supported encoding and compression schemes. I suggest you change how you're recording the data, or introduce another step in your pipeline to first uncompress the audio data before uploading it.

Closing this issue now as both the library and service are working as expected.

**Comment by harsh-tukadiya** [comment]
@jskeet 
Thank you very much for helping  :)

**Comment by harsh-tukadiya** [comment]
@jskeet 
In this all file i convert speech to text.
But they need **Audiochannelcount = 2 ** and sometimes **audiochannelcount = 1**
So, what can  i do for this issue.

**Comment by jskeet** [comment]
@harsh-tukadiya: I'm afraid I don't understand your question, but you should file a new issue if you absolutely need to (but only after trying to act on the rest of this comment). You should also try not specifying `AudioChannelCount` at all - I *suspect* it will be autodetected for you. If not, you should probably read the audio file header yourself so you can specify the right number of channels.

**Comment by shreyshahh** [comment]
Hi @jskeet, it is not detecting AudioChannelCount automatically for me, how can I dynamically give that count? How can  I read audio file header programmatically?

**Comment by jskeet** [comment]
@shreyshahh: Please don't use existing issues to raise new concerns. But I would suggest using [NAudio](https://github.com/naudio/NAudio)

**Comment by shreyshahh** [comment]
okay, thanks. I'll create new issues from now ;)

**Comment by smarty5** [comment]
You can try this [speech to text](https://audext.com/wav-to-text/) converter that supports .wav format. It has 30 minutes for free and works very well. If it's necessary to keep your data confidential, I recommend using any paid converter that won't disclose information due to their privacy policy.

**Comment by jskeet** [comment]
(Comment from another user deleted as effectively an irrelevant advert for a separate product.)
